### ISSUE TITLE
@Template.directives can take Bindings (not just types)

### ISSUE THREAD

**Issue by mhevery** [angular/angular#709]
The `@Template.directives` should take full [`Binding`](https://github.com/angular/angular/blob/821f008462abdb24a18d4852bfe3fe04d8695823/modules/angular2/src/di/binding.js#L21) support.

```
@Template({
  url: '...',
  directives: [
    Directive1, 
    bind(Directive2).toClass(Directive2Impl),
  ]
}) 
```

## UPDATE 1

The reason for this issue is something like this.

```
<tabs>
  <pane></pane>
</tabs>
```

```
@Component({selector: 'tabs'})
class Tabs {}

@Component({selector: 'pane'})
class Pane {
  constructor(tabs:Tabs) {}
}
```

Here `Pane` can inject `Tabs`. But what if you would like to write a different tab which participates is this. Imagine this.

```
<view-stack>
  <pane></pane>
</view-stack>
```

```
@Component({selector: 'view-stack'})
class ViewStack extends Tabs {}
```

The above would not work even thought `ViewStack` implements `Tabs`.

So to fix this we would do this:

```
  directives: [
    Pane, 
    bind(Tabs).toClass(ViewStack),
  ]
```

And it would work as expected.

**Comment by SekibOmazic** [comment]
Hi ng-team,

I tried to solve this issue and came up with the following solution:
https://github.com/SekibOmazic/angular/commit/bfb3acc876bbdcdf386b895b4679e4e5ec836c64

Since I'm not sure if this is good enough, so I haven't made a PR yet. It would be nice if someone could look at this and give me feedback. If this is OK I'd send a PR.

Thanks!

**Comment by vicb** [comment]
@SekibOmazic It would be much simpler to discuss on a PR, do not hesitate to send (and don't be afraid to be wrong, you'll learn).

` Object.getPrototypeOf(item.factory()).constructor`  is probably problematic:
- It would not work with Dart target,
- It would not work with `bind(...).toValue()` or `bind(...).toFactory()`

Do we actually need to special case `Binding` ?

The following code might help: https://github.com/angular/angular/blob/master/modules/angular2/src/di/injector.js#L247

**Comment by vicb** [comment]
I just had a quick look at the code and it looks like the implementation would require more changes to the compiler internals.

@mhevery when we have something like  `bind(Directive2).toClass(Directive2Impl)`, should the annotation be retrieved from `Directive2` or `Directive2Impl` ?
_If we want to be able to have any Binding, `bind(Directive2).toValue(directive)` should also be supported_

**Comment by SekibOmazic** [comment]
@vicb Thanks for the feedback! 
I've just tried another test with: `bind(SomeDirective).toValue(new SomeDirective())` and that worked (It still has to be thoroughly tested, of course).

Could you also give a hint, where to look in Dart code? (Please bear with me, I'm Dart newbe)

**Comment by vicb** [comment]
The same code is transpiled to either JS or Dart. However you should make sure that your code work for both. 

You should run the tests for both version `gulp test.unit.js` & `gulp test.unit.dart`.

Check modules/angular2/src/facade to see how you can execute dart or js specific code according to the target language. ie lang.es6 and lang.dart.

**Comment by pkozlowski-opensource** [comment]
@vicb actually, U was wondering, what was the motivation behind this issue? I mean, directives got a selector that will match occurrences in templates...

**Comment by SekibOmazic** [comment]
@vicb It really seems that a bigger change to the compiler is needed. Dart test failed with:
```Shell
expression is not a valid compile-time constant
@Template(inline: '<p>Component with an injected parent</p>', directives: const[bind(SomeDirective).toClass(SomeDirective)]) 
                                                                                ^
```

Well, at least I learned more about Angular compiler.

**Comment by mhevery** [comment]
@SekibOmazic updated the issue with reasoning.

**Comment by SekibOmazic** [comment]
@mhevery Thanks.

**Comment by matanlurey** [comment]
The syntax needs to be slightly different to support Dart static analysis, e.g.:

```dart
directives: const [
  Pane, 
  const Bind(Tabs, toClass: ViewStack)
]
```

**Comment by naomiblack** [comment]
@pkozlowski-opensource do you have time to take this one?

**Comment by ajoslin** [comment]
We need this in Ionic.

@pkozlowski-opensource I could try my best to implement this if you point me in the right direction.

**Comment by pkozlowski-opensource** [comment]
@ajoslin unfortunately, looking through the comments, it might be blocked by a Dart transpilation issue. @rkirov isn't this the same issue with `new` being generated instead of `const`? 

@ajoslin let me confirm that there is a transpilation issue here, will look into this one today / tomorrow.

**Comment by matanlurey** [comment]
@pkozlowski-opensource: Bindings must be made entirely by constants in Dart, i.e.:

```dart
services: const [
  const Bind(Foo, toClass: FooImpl)
]
```

not

```js
services: [
  bind(Foo).toClass(FooImpl)
]
```

**Comment by pkozlowski-opensource** [comment]
Here is a minimal reproduce scenario, just to make sure that I understand the use-case correctly:
http://plnkr.co/edit/gAWUXXs0XUURwW0OxRq7?p=preview

**Comment by ajoslin** [comment]
@pkozlowski-opensource You look to be on the right track. Just to be sure, I changed it to my exact usecase: http://plnkr.co/edit/8LSDpr7zIqhpGF3o7Z87?p=preview

I have a generic 'ViewStack', which is a normal container for content. And then each pane in a 'tabs' view is a subclass of 'ViewStack'.

If, in this example, the Pane is able to inject the 'ViewStack' ancestor whether the ancestor is a tabs or a view stack, all is well.

in that example, I'd want 'pane' to inherit

**Comment by pkozlowski-opensource** [comment]
@ajoslin right, we do talk about the same case. Here is further reduction: http://plnkr.co/edit/b5R775RTAvaIWzjCAp6j?p=preview

Diving into ElementInjector....

**Comment by yjbanov** [comment]
We need to change our binding API before tackling this issue. The existing `bind(Blah).toClass(Bleu)` isn't producing constants. @mhevery Could you update the description to be more accurate with requirements? I think the implementation is a two-step process:

1. Refactor the binding API so it expresses bindings via constants
2. Implement @Template.directives

It doesn't make sense to tackle the second problem when the binding API is not functional.

**Comment by yjbanov** [comment]
I logged the prerequisite for this as #1229

**Comment by tbosch** [comment]
Note that the current compiler refactorings that I am doing path the way
for this already, as internally, the Compiler already uses
DirectiveBindings...

**Comment by matanlurey** [comment]
Is this closer to being  implementable/implemented @tbosch?

**Comment by matanlurey** [comment]
Also, how would `@DynamicComponent` participate in this?

**Comment by tbosch** [comment]
@matanlurey This should work with a little change in [compiler.js](https://github.com/angular/angular/blob/master/modules/angular2/src/core/compiler/compiler.js#L76): 

```
  _bindDirective(directiveTypeOrBinding) {
    if (directiveTypeOrBinding instanceof DirectiveBinding) {
      return directiveTypeOrBinding;
    } else if (directiveTypeOrBinding instanceof Binding) {
      var meta = this._reader.read(directiveTypeOrBinding.token);
      return DirectiveBinding.createFromBinding(directiveTypeOrBinding, meta.annotation);
    } else {
      var meta = this._reader.read(directiveTypeOrBinding);
      return DirectiveBinding.createFromType(meta.type, meta.annotation);
    }
  }
```

**Comment by matanlurey** [comment]
:+1:

**Comment by yjbanov** [comment]
Should we close this issue or is there more to be done?

**Comment by pkozlowski-opensource** [comment]
@yjbanov the original issue reported by @mhevery is solved (by #1498) but I don't think it is enough for other cases brought by component authors (see @ajoslin use-case in this issue). My concerns were expressed here: https://github.com/angular/angular/pull/1498#issuecomment-95931839

In short: I would need someone's input on the @ajoslin use-case - that is - do you guys think it is valid and how we should approach it if it is valid. We can do it in this or a separate issue, I don't mind.

**Comment by ajoslin** [comment]
@pkozlowski-opensource I may not have clearly understood the code.

But I don't think this quite covers our use case.

I need the child to be able to inject the `GenericParent` without knowledge of how it's implemented.

```js
@Component()
class NavigationStack {} //push/pop/animate between views

@Component()
class Tab extends Parent {} //push/pop/animate between views and do tab-things

// User's component
@Component()
class MyComponent {
  constructor(navStack: NavigationStack) {
    // I don't need to know if navStack is a `Tab` or not.
  }
}
```

**Comment by pkozlowski-opensource** [comment]
@ajoslin this is what I thought... I would propose that we open a new issue with the latest plunk as the discussion here touches upon several points and we might be loosing focus here.

I'm going to close this one, could you open a new issue with a latest plunk and a minimal use-case? Thnx!

**Comment by ajoslin** [comment]
Yes.
